zk/datastream/client/stream_client.go

```diff
@@ -86,6 +95,93 @@ func (c *StreamClient) IsVersion3() bool {
 func (c *StreamClient) GetEntryChan() chan interface{} {
 	return c.entryChan
 }
+
+func (c *StreamClient) GetL2BlockByNumber(blockNum uint64) (*types.FullL2Block, error) {
+	if _, err := c.EnsureConnected(); err != nil {
+		return nil, err
+	}
+	defer c.Stop()
+
+	var (
+		l2Block   *types.FullL2Block
+		err       error
+		isL2Block bool
+	)
+
+	bookmark := types.NewBookmarkProto(blockNum, datastream.BookmarkType_BOOKMARK_TYPE_L2_BLOCK)
+	bookmarkRaw, err := bookmark.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.initiateDownloadBookmark(bookmarkRaw); err != nil {
+		return nil, err
+	}
+
+	for l2Block == nil {
+		select {
+		case <-c.ctx.Done():
+			return l2Block, nil
+		default:
+		}
+
+		parsedEntry, err := c.readParsedProto()
+		if err != nil {
+			return nil, err
+		}
+
+		l2Block, isL2Block = parsedEntry.(*types.FullL2Block)
+		if isL2Block {
+			break
+		}
+	}
+
+	if l2Block.L2BlockNumber != blockNum {
+		return nil, fmt.Errorf("expected block number %d but got %d", blockNum, l2Block.L2BlockNumber)
+	}
+
+	return l2Block, nil
+}
+
+func (c *StreamClient) GetLatestL2Block() (l2Block *types.FullL2Block, err error) {
```

Comment:
Same as above. Unit test would be great.

Reply:
Same as in the https://github.com/0xPolygonHermez/cdk-erigon/pull/886#discussion_r1719886959. Tricky to test, because function uses the same connection to write and read, which is blocking.

Reply:
After playing around a bit more with `net.Pipe` connection, I've managed to implement a unit test: https://github.com/0xPolygonHermez/cdk-erigon/blob/6b0c1589fba196858efbd54ca13177d70c354768/zk/datastream/client/stream_client_test.go#L272-L410. If you agree I'm about to do the same for `GetL2BlockByNumber`.

Reply:
This looks great. Thanks! Yes, please have a test for `GetL2BlockByNumber` if possible.